src/util/bson.hpp

```diff
@@ -0,0 +1,1218 @@
+            case state_t::object_id:
+                m_marks.top().push_back({instruction.key, ObjectId(val.data())});
+                m_instructions.push({state_t::skip});
+                break;
+            case state_t::date:
+                m_marks.top().push_back({instruction.key, time_t(atol(val.data()))});
+                // skip twice because this is a number long
+                m_instructions.push({state_t::skip});
+                m_instructions.push({state_t::skip});
+                break;
+            case state_t::regular_expression_pattern:
+                // if we have already pushed a regex type
+                if (m_marks.top().size() && m_marks.top().back().first == instruction.key) {
+                    auto regex = std::get<RegularExpression>(m_marks.top().back().second);
+                    m_marks.top().pop_back();
+                    m_marks.top().push_back({instruction.key, RegularExpression(val, regex.options())});
+
+                    // pop vestigal regex instruction
+                    m_instructions.pop();
+                    m_instructions.push({state_t::skip});
+                    m_instructions.push({state_t::skip});
+                } else {
+                    m_marks.top().push_back({instruction.key, RegularExpression(val, "")});
+                }
+
+                break;
+            case state_t::regular_expression_options:
+                // if we have already pushed a regex type
+                if (m_marks.top().size() && m_marks.top().back().first == instruction.key) {
+                    auto regex = std::get<RegularExpression>(m_marks.top().back().second);
+                    m_marks.top().pop_back();
+                    m_marks.top().push_back({instruction.key, RegularExpression(regex.pattern(), val)});
+                    // pop vestigal regex instruction
+                    m_instructions.pop();
+                    m_instructions.push({state_t::skip});
+                    m_instructions.push({state_t::skip});
+                } else {
+                    m_marks.top().push_back({instruction.key, RegularExpression("", val)});
+                }
+
+                break;
+            case state_t::binary_sub_type:
+                // if we have already pushed a binary type
+                if (m_marks.top().size() && m_marks.top().back().first == instruction.key) {
+                    // we will ignore the subtype for now
+                    // pop vestigal binary instruction
+                    m_instructions.pop();
+                    m_instructions.push({state_t::skip});
+                    m_instructions.push({state_t::skip});
+                } else {
+                    // we will ignore the subtype for now
+                    m_marks.top().push_back({instruction.key, BinaryData()});
+                }
+
+                break;
+            case state_t::binary_base64: {
+                char* binary_val = (char *)malloc(sizeof(char) * val.size());
```

Comment:
Writing `malloc` is basically always incorrect, and this is no exception. This memory is simply leaked currently.